platform-sdk/docs/components/componentFramework.md

```diff
@@ -0,0 +1,200 @@
+
+Component 5 is a Concurrent task scheduler which uses the `ForkJoinPool` to execute tasks. Each task is executed by some
+FJP thread. The thread that executes a Component 5 task sends the resulting data to Component 6 and creates and
+schedules the Component 6 task. This task may be executed on the same FJP thread or a different one.
+
+Component 7 is another Sequential Thread task scheduler, which supports multiple threads scheduling tasks
+simultaneously. Component 7's tasks are executed by a dedicated thread, and the resulting data is used to create and
+execute Component 8 tasks since Component 8 is a Direct task scheduler that executes tasks immediately on the scheduling
+thread.
+
+## Building a Pipeline
+
+The steps below describe how simple components can be created and wired together. This is one example of the order of
+steps.
+
+### Step 1 - Create the Task Scheduler
+
+Create the `TaskScheduler` parameterized with the type of data the component will produce. The `TaskScheduler` comes
+with an `OutputWire` build in.
+
+```
+TaskScheduler<String> fooTaskScheduler = WiringModel.schedulerBuilder("Foo");
+```
+
+![Step 1](step1.drawio.svg)
+
+### Step 2 - Create the Input Wire
+
+Create the InputWire for the TaskScheduler with the type of data the component will accept.
+
+```
+InputWire<Integer> fooInputWire = fooTaskScheduler.buildInputWire("Foo Input");
+```
+
+![Step 2](step2.drawio.svg)
+
+### Step 3 - Create the Business Logic
+
+Create the business logic required to operate on the input data and produce the output data. At this
+point, the component is completely disconnected from the TaskScheduler.
+
+```
+BusinessLogic logic = new BusinessLogic();
+```
+
+![Step 3](step3.drawio.svg)
+
+### Step 4 - Bind the Business Logic
+
+Connect the business logic to the TaskScheduler by binding the InputWire to the method that operates on the input data
+and produces the output data. The TaskScheduler is now connected, and will automatically forward the data produced by
+the bound method to the OutputWire. The component is now built.
+
+```
+fooInputWire.bind(logic::intToString);
+```
+
+![Step 4](step4.drawio.svg)
+
+### Step 5 - Connecting Components
+
+Connect the component to other components by soldering input and output wires.
+
+```
+fooTaskScheduler.getOutputWire().solderTo(barInputWire);
+barTaskScheduler.getOutputWire().solderTo(bazInputWire);
+```
+
+![Step 5](step5.drawio.svg)
+
+## Features
+
+### Wiring Model
+
+The wiring model tracks all task schedulers and wires that connect them and is capable of doing useful analysis of the
+resulting graph. Most notably, it can produce mermaid style diagrams that show the flow of data through wires and
+components, detect cyclic backpressure that could result in a deadlock, and verify proper wiring of various task
+schedulers according to their restrictions (for example, a concurrent task scheduler is not permitted to be wired to a
+direct scheduler).
+
+### Wire Add Ons
```

Comment:
BTW, not directly related to this PR, but if feels like each of the add ons listed below could just be a component with a business logic doing its thing and using a Direct task scheduler to avoid any scheduling overhead. Are there any advantages of introducing a new concept of an add-on? Would it be simpler to only have components in the framework?

Reply:
The goal of add ons was to take a pattern that we expect to see repeated in many places, and to write boilerplate code to do it make life easier for an end user. Each of these add ons could in theory be emulated entirely by non-framework code, but do do so would result in lots of repeated code.

Reply:
I'm not sure about the "lots of repeated code" though. A generic component implementing, say, the filter could be parametrized with a data type-specific Function<>. This same generic component implementation could also expose a public static method to simplify adding this component to the data pipeline.

This way the data pipeline configuration code would only need to implement that single method in the Function<> (which, I presume, it has to do today anyway - and if there are repeated use-cases, then the implementation could obviously be extracted into some common package), and then add a single call to that static method to actually add this "add-on" component.

This way no code would ever be repeated. Yet the framework would be simplified because of the less number of concepts/entities in its architecture.

Reply:
I see the merits of both approaches. An add on is nothing more than a direct scheduler that happens to have very simple business logic, so why not make it just that? But I think I lean toward keeping them as is because conceptually, a component as an essential part of the system that performs critical business logic. Something that has a business reason for existing. In my opinion, the logic of add ons does not rise to the level. Their operations are trivial and more of a wire compatibility issue than core business logic, therefore they should not be components. That being said, I am not adamant one way or another.